src/semant/translate.sml

```diff
@@ -270,4 +272,28 @@ structure Translate : TRANSLATE = struct
     in
       frags := Frame.PROC(proc)::(!frags)
     end
+
+  fun record(fields) =
+    let val size = T.TEMP(Temp.newtemp())
+        val ret = T.TEMP(Temp.newtemp())
+        val alloc = T.BINOP(T.MUL, T.CONST(Frame.wordSize), size)
+        val offset = ref 0
+        fun init(e) = (offset := !offset + Frame.wordSize;
+                       T.MOVE(T.BINOP(T.PLUS,ret,T.CONST(!offset)), unEx(e)))
+        val init_values = map init fields
+        val setup = [T.MOVE(size, T.CONST(List.length(fields))),
+                     T.MOVE(ret, Frame.externalCall("malloc", [alloc]))]
+    in
+      Ex(T.ESEQ(seq(setup @ init_values), ret))
+    end
+
+  fun fieldVar(v,s,l) =
+    let fun getOffset(_,nil) = raise TypeCheckFailed
```

Comment:
Not sure this is a type error, but rather something we expect to be true since the program is assumed to be well typed.